src/radiation_system.hpp

```diff
@@ -917,189 +1022,244 @@ void RadSystem<problem_t>::AddSourceTerms(array_t &consVar, arrayconst_t &radEne
+			consNew(i, j, k, x1RadFlux_index + numRadVars_ * g) = Frad_t1[0];
+			consNew(i, j, k, x2RadFlux_index + numRadVars_ * g) = Frad_t1[1];
+			consNew(i, j, k, x3RadFlux_index + numRadVars_ * g) = Frad_t1[2];
 
-		consNew(i, j, k, x1GasMomentum_index) += dMomentum[0];
-		consNew(i, j, k, x2GasMomentum_index) += dMomentum[1];
-		consNew(i, j, k, x3GasMomentum_index) += dMomentum[2];
+			// compute the gas momentum after adding the current group's momentum
+			consNew(i, j, k, x1GasMomentum_index) += dMomentum[0];
+			consNew(i, j, k, x2GasMomentum_index) += dMomentum[1];
+			consNew(i, j, k, x3GasMomentum_index) += dMomentum[2];
+		}
 
+		// update gas momentum from radiation momentum of the current group
 		if constexpr (gamma_ != 1.0) {
 			amrex::Real x1GasMom1 = consNew(i, j, k, x1GasMomentum_index);
 			amrex::Real x2GasMom1 = consNew(i, j, k, x2GasMomentum_index);
 			amrex::Real x3GasMom1 = consNew(i, j, k, x3GasMomentum_index);
 
 			// 4a. Compute radiation work terms
 			amrex::Real const Egastot1 = ComputeEgasFromEint(rho, x1GasMom1, x2GasMom1, x3GasMom1, Egas_guess);
-			amrex::Real dErad_work = NAN;
+			amrex::Real dErad_work = 0.;
 
 			if constexpr (compute_v_over_c_terms_) {
-				// compute difference in gas kinetic energy before and after momentum
-				// update
+				// compute difference in gas kinetic energy before and after momentum update
 				amrex::Real const Ekin1 = Egastot1 - Egas_guess;
 				amrex::Real const dEkin_work = Ekin1 - Ekin0;
 				// compute loss of radiation energy to gas kinetic energy
 				dErad_work = -(c_hat_ / c_light_) * dEkin_work;
+
+				// compute energyLossFractions
+				quokka::valarray<double, nGroups_> energyLossFractions{};
+				for (int g = 0; g < nGroups_; ++g) {
+					Frad_t1[0] = consNew(i, j, k, x1RadFlux_index + numRadVars_ * g);
+					Frad_t1[1] = consNew(i, j, k, x2RadFlux_index + numRadVars_ * g);
+					Frad_t1[2] = consNew(i, j, k, x3RadFlux_index + numRadVars_ * g);
+					energyLossFractions[g] = kappaFVec[g] * (x1GasMom1 * Frad_t1[0] + x2GasMom1 * Frad_t1[1] + x3GasMom1 * Frad_t1[2]);
+				}
+				// TODO(CCH): tricky here. Have to avoid negavie radiation energy.
+				auto energyLossFractionsTot = sum(energyLossFractions);
+				if (energyLossFractionsTot != 0.0) {
+					energyLossFractions /= energyLossFractionsTot;
+				} else {
+					energyLossFractions.fillin(0.0);
+				}
+				for (int g = 0; g < nGroups_; ++g) {
+					auto radEnergyNew = EradVec_guess[g] + dErad_work * energyLossFractions[g];
+					consNew(i, j, k, radEnergy_index + numRadVars_ * g) = radEnergyNew;
+				}
 			} else {
-				// do not subtract radiation work in new radiation energy
-				dErad_work = 0.;
+				// else, do not subtract radiation work in new radiation energy
+				for (int g = 0; g < nGroups_; ++g) {
+					consNew(i, j, k, radEnergy_index + numRadVars_ * g) = EradVec_guess[g];
+				}
 			}
 
-			// compute difference between new and old internal energy
-			amrex::Real const dEint = Egas_guess - Egas0;
```

Comment:
Is there a reason to get rid of this variable?

Reply:
Not really. I like to keep the code concise and avoid unnecessary variables, but I also appreciate extra variables to better explain the physics. I put dEint back.